jupyterhub/singleuser/extension.py

```diff
@@ -0,0 +1,659 @@
+"""
+Integrate JupyterHub auth with Jupyter Server as an Extension
+
+Requires Jupyter Server 2.0, which in turn requires Python 3.7
+"""
+
+from __future__ import annotations
+
+import asyncio
+import json
+import os
+import random
+from datetime import timezone
+from functools import wraps
+from pathlib import Path
+from unittest import mock
+from urllib.parse import urlparse
+
+from jupyter_core import paths
+from jupyter_server.auth import Authorizer, IdentityProvider, User
+from jupyter_server.auth.logout import LogoutHandler
+from jupyter_server.extension.application import ExtensionApp
+from tornado.httpclient import AsyncHTTPClient, HTTPRequest
+from tornado.httputil import url_concat
+from tornado.web import HTTPError
+from traitlets import Any, Bool, Instance, Integer, Unicode, default
+
+from jupyterhub._version import __version__, _check_version
+from jupyterhub.services.auth import HubOAuth, HubOAuthCallbackHandler
+from jupyterhub.utils import (
+    exponential_backoff,
+    isoformat,
+    make_ssl_context,
+    url_path_join,
+)
+
+SINGLEUSER_TEMPLATES_DIR = str(Path(__file__).parent.resolve().joinpath("templates"))
+
+
+def _bool_env(key):
+    """Cast an environment variable to bool
+
+    0, empty, or unset is False; All other values are True.
+    """
+    if os.environ.get(key, "") in {"", "0"}:
+        return False
+    else:
+        return True
+
+
+def _exclude_home(path_list):
+    """Filter out any entries in a path list that are in my home directory.
+
+    Used to disable per-user configuration.
+    """
+    home = os.path.expanduser('~/')
+    for p in path_list:
+        if not p.startswith(home):
+            yield p
+
+
+class JupyterHubLogoutHandler(LogoutHandler):
+    def get(self):
+        hub_auth = self.identity_provider.hub_auth
+        # clear single-user cookie
+        hub_auth.clear_cookie(self)
+        # redirect to hub to clear the rest
+        self.redirect(hub_auth.hub_host + url_path_join(hub_auth.hub_prefix, "logout"))
```

Comment:
I've added a small clarification that `hub_auth` is responsible for the cookie. I don't like too many links to specific docs sections from code because they easily go stale and aren't ever checked. I think keywords and description that can be fed to search are more durable and reliable.